[Property wrappers] Allow default initialization to infer a type.

DougGregor opened the pull request:
Property wrappers are allowed to infer the type of a variable, but this
only worked when the property wrapper was provided with an explicit
initialization, e.g.,

    @WrapsAnInt() var x   // infers type Int from WrapsAnInt.wrappedValue

However, when default initialization is supported by the property wrapper,
dropping the parentheses would produce an error about the missing type
annotation

    @WrapsAnInt var x

Make this second case behave like the first, so that default initialization
works consistently with the explicitly-specified version.
Fixes rdar://problem/59471019.

DougGregor committed:
[Type checker] Remove dead code to compute pattern checking options.

DougGregor committed:
[Property wrappers] Allow default initialization to infer a type.

Property wrappers are allowed to infer the type of a variable, but this
only worked when the property wrapper was provided with an explicit
initialization, e.g.,

    @WrapsAnInt() var x   // infers type Int from WrapsAnInt.wrappedValue

However, when default initialization is supported by the property wrapper,
dropping the parentheses would produce an error about the missing type
annotation

    @WrapsAnInt var x

Make this second case behave like the first, so that default initialization
works consistently with the explicitly-specified version.
Fixes rdar://problem/59471019.

DougGregor commented:
@swift-ci please smoke test

DougGregor commented:
@swift-ci please test source compatibility

hborla commented on lib/AST/Decl.cpp:
```diff
@@ -1864,13 +1875,13 @@ bool PatternBindingDecl::isDefaultInitializable(unsigned i) const {
+  if (isDefaultInitializableViaPropertyWrapper(i))
+    return true;
+
+  // If one of the attached wrappers is missing an initialValue
```
Nit: should we be using `wrappedValue` terminology from now on? All the appearances of `initialValue` confused me at first because I didn't realize the name had changed over the course of review

DougGregor replied:
Yes, you're right. I copied this code from elsewhere and missed that this was out-of-date, thanks!

DougGregor replied:
Looks like there are a couple of APIs that also mention `initialValue`, so there's more cleanup to do here.

DougGregor commented:
The Release source compatibility build failed with:

DougGregor committed:
[Property wrappers] NFC rename initialValue -> wrappedValue where appropriate

wrappedValue is the new name, and we only support initialValue for backward
compatibility with earlier revisions of the property wrappers proposal.

xymus commented:
Let's see if the source compat bots are in a better shape now.

@swift-ci please test source compatibility

DougGregor commented:
@swift-ci please smoke test

DougGregor commented:
Source compatibility bots were fine; I've pulled in the cleanup @hborla suggested and will land with that.

DougGregor commented:
@swift-ci please smoke test
